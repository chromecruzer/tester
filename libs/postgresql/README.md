# @trac-postgresql

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test @trac-postgresql` to execute the unit tests via [Jest](https://jestjs.io).
