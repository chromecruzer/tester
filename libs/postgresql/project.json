{
  "name": "postgresql",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/postgresql/src",
  "projectType": "library",
  "targets": {
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/postgresql/**/*.ts"]
      }
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["{workspaceRoot}/coverage/libs/postgresql"],
      "options": {
        "jestConfig": "libs/postgresql/jest.config.ts",
        "passWithNoTests": true
      }
    }
  },
  "tags": []
}
