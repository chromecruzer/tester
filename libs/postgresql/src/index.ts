export * from './postgresql';
import PostgresConfigFactory, {PostgresConfig} from "./PostgresConfigFactory";
import BulkSqlWithCreate from "./BulkSqlWithCreate";
export *  from './ErrorHandler';
import ExtractIolData from "./ExtractIolData";
import PostgresClient from "./PostgresClient";
import LoadXlsxData from './LoadXlsxData';
import NotificationsServer from "./NotificationsServer";
import SearchTable from "./SearchTable";
import ImportIntoConsignments from "./ImportIntoConsignments";
import ImportIntoJumpTable from "./ImportIntoJumpTable";
import AuditStore from "./AuditStore";
import UserStore from './UserStore';

export {
  AuditStore, PostgresConfig, PostgresConfigFactory, BulkSqlWithCreate, ExtractIolData,
  PostgresClient, LoadXlsxData, NotificationsServer, SearchTable, ImportIntoConsignments, ImportIntoJumpTable,
  UserStore
};
