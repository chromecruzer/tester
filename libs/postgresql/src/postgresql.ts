import * as _ from "lodash";
import {DateTime} from "luxon";
import {PoolClient} from "pg";
import util from "util";
// eslint-disable-next-line @nrwl/nx/enforce-module-boundaries
import {
  customerContactDataMapping,
  customerDataMapping,
  dataTableNames, InputDataType,
  InputDataTypeLabel,
  InputToSqlMapping,
  iolDataMapping,
  matchCustomerContactFields,
  matchCustomerFields,
  matchIolFields,
  matchProductFields,
  matchSalesMappingFields,
  matchSalesRepFields,
  NullableString,
  productDataMapping,
  employeeRecordDm,
  salesMappingDataMapping,
  salesRepDataMapping,
  tracDateFormat, 
  UploadRecord,
  matchEmployeeFields,
  userRecordDm,
  matchUser1Fields
} from "../../datatypes/src";
import path from "path";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export const auditTableNames = {
  notes: 'auditnotes',
  locations: 'auditlocation',
  items: 'audititem',
}



export const searchTableNamesFn = () => [
  dataTableNames.consignments,
  dataTableNames.customers,
  dataTableNames.customerContacts,
  dataTableNames.products,
  dataTableNames.salesReps];

export const sqlDate = (datum: (Date | string)) => {
  // console.log(`datum ${dump(datum)}
  // from JSDate ${dump(DateTime.fromJSDate(datum))}
  // to format ${dump(DateTime.fromJSDate(datum).toFormat(tracDateFormat))}`)
  const date = _.isDate(datum) ? datum : new Date(datum);
  return DateTime.fromJSDate(date).toFormat(tracDateFormat)
}

export function datumToSql(datum: (string | number | Date | boolean), datatype: InputDataTypeLabel) {
  if (_.isNull(datum) || _.isUndefined(datum) || datum === '') {
    return 'NULL';
  }
  let str;
  switch (datatype) {
    case 'string':
      str = String(datum);
      if (str.length === 0) {
        return null;
      }
      str = str.replace(/'/g, "''");
      str = str.replace(/\n/g, "\\n");
      return `'${str}'`;
    case 'date':
      // console.log(`date ${datum} converted ${sqlDate(datum as unknown as Date)}`)
      return `'${sqlDate(datum as Date)}'`
    case 'boolean':
      // console.log(`boolean ${datum} as ${Boolean(datum)}`)
      return Boolean(datum);
    case 'number':
      return datum;
  }
}

export function appendUuids(sql: string, uuids: string[]) {
  let skipped1st = false;
  let result = sql;
  uuids.forEach(id => {
    result += `${skipped1st ? ',' : ''}    \n'${id}'`;
    skipped1st = true;
  });
  result += ');';
  return result;
}

export const checkModified = modified => {
  // console.log(`Checking ${dump(modified)}`);
  if (modified.originalUuid !== modified.new.uuid) {
    console.error(`UUID mismatch found ${dump(modified)}`);
  }
}

export function appendBulkData(sql, bulkData: UploadRecord[],
                               columnMapping: InputToSqlMapping[],
                               createUuid = true,
                               hasUuid = false) {
  let skipped1stRow = false;
  let result = `${sql}\n`;
  bulkData.forEach((record) => {
    // console.log(`append bulk data ${dump(record)}`)
    const dataRow = [];
    if (createUuid) {
      dataRow.push('gen_random_uuid()')
    }
    if(hasUuid) {
      dataRow.push(datumToSql(record['uuid'], 'string'));
    }
    columnMapping.forEach((colMap) => {
      let datum = datumToSql(record[colMap.sqlLabel], colMap.dataType);
      if(!_.isUndefined(colMap.default) && datum === null) {
        datum = colMap.default;
      }
      dataRow.push(datum)
    });
    result += `${skipped1stRow ? ',': ''}\n(${_.join(dataRow, ', ')})`;
    skipped1stRow = true;
  });
  return result;
}

export const queryRecords = async (client: PoolClient, recordSql: string, idField: string,
                                   {
                                     uuid = null as NullableString,
                                     id = null as NullableString,
                                     ids = [] as string[],
                                   }) => {
  let sql = recordSql;
  const check = (_.isString(uuid) ? 1 : 0) + (_.isString(id) ? 1 : 0) + (_.isEmpty(ids) ? 0 : 1);
  if (check > 1) {
    throw new Error(`cannot set multiple parameters ${uuid}, ${id}, ${ids.length}`);
  }
  if (uuid) {
    sql += ` WHERE uuid='${uuid}'`;
  }
  if (id) {
    sql += ` WHERE ${idField}='${id}'`;
  }
  if (!_.isEmpty(ids)) {
    sql += ` WHERE ${idField} IN (${_.join(_.map(ids, i => `'${i}'`), ', ')})`;
  }
  sql += ';';
  // console.log(`query records sql '${sql}'`);
  return client.query(sql)
    .then(data => data.rows)
    .catch(err => {
    throw new Error(`Mistake in '${sql}' caused ${err.message}`);
  });
}

interface TransformerKeys {
  matchFields: (string | string[]);
  matchFn: (a: any, b: any) => boolean,
  mappings: InputToSqlMapping[];
  compareFieldList?: string[],
}

export const recordFromArray = (row: InputDataType[], mapping: InputToSqlMapping[]) => {
  return _.reduce(mapping, (accum, m,i) => {
    const r = row[i] || null;
    let value = _.isFunction(m.nullify) ? m.nullify(r) : r;
    if(_.has(m, 'default')) { //need to us _.has here because the default may be a boolean false
      // console.log(`checking default for ${dump(m)} with value ${value}`)
      value = value || m.default;
    }
    accum[m.sqlLabel] = value;
    return accum;
  }, {} as UploadRecord);
}
export const transformHelpers: { [id: string]: TransformerKeys } = {
  iol: {
    matchFields: 'lot',
    matchFn: matchIolFields,
    mappings: iolDataMapping,
  },
  customers: {
    matchFields: 'customer_code',
    matchFn: matchCustomerFields,
    mappings: customerDataMapping,
  },
  customer_contacts: {
    matchFields: ['full_name', 'ship_to_id'],
    matchFn: matchCustomerContactFields,
    mappings: customerContactDataMapping,
  },
  products: {
    matchFields: 'product_id',
    matchFn: matchProductFields,
    mappings: productDataMapping,
  },
  employee: {
     matchFields: 'emp_id',
     matchFn: matchEmployeeFields,          // added by shyam for testing
     mappings: employeeRecordDm,
  },

  users: {
    matchFields: 'username',
    matchFn:  matchUser1Fields,              // added by shyam for testing
    mappings: userRecordDm
  },
  salesmappings: {
    matchFields: 'customer_ship_to',
    matchFn: matchSalesMappingFields,
    mappings: salesMappingDataMapping,
    compareFieldList: ['customer_ship_to', 'postal', 'zipcode', 'primary_territory', 'primary_name', 'cos_territory',
      'cos_name', 'kae_territory', 'kae_name'],
  },
  salesreps: {
    matchFields: 'salesrep_id',
    matchFn: matchSalesRepFields,
    mappings: salesRepDataMapping,
  },
}
export const matchFieldFn = (fields, record) => {
  return _.isArray(fields) ? _.join(_.map(fields, f => record[f]), '|') : record[fields]
}
export const deepEqualFn = (c, m, compareFieldList = null) => {
  if (_.isArray(compareFieldList)) {
    return _.reduce(compareFieldList,
      (boolAccum, cf) => {
        if (!boolAccum) {
          return boolAccum;
        }
        // console.log(`comparing isequal field ${cf}, ${candidate[cf]} to ${m[cf]}`);
        return _.isEqual(c[cf], m[cf]);
      }, true);
  }
  return _.isEqual(m, c);
}
export const docxTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.docx' : 'expiredTemplate.docx'))
export const htmlTemplateFn = (apiConfig, templateType) => path.join(apiConfig.templatesPath,
  (templateType === 'Missing' ? 'missingTemplate.html' : 'expiredTemplate.html'))

