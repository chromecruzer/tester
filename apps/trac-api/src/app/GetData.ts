import {PostgresClient, PostgresConfigFactory, queryRecords, SearchTable, tryCatch} from "@trac/postgresql";
import {getDataFields} from "@trac/datatypes";
import util from "util";

const dump = (obj, depth = null) => util.inspect(obj, {depth, colors: true});

export default class GetData {
  searchTable: SearchTable;
  getConsignmentsSql: string;
  getCustomersSql: string;
  private getCustomerContactsSql: string;
  private getSalesMappingsSql: string;
  constructor(private postgresConfig: PostgresConfigFactory, consignmentTable: string, customersTable: string,
              customerContactTable: string, salesMappingsTable: string) {
    this.searchTable = new SearchTable(postgresConfig.getSchemaName());
    this.getCustomersSql = `SELECT * FROM ${postgresConfig.getSchemaName()}.${customersTable}`;
    this.getCustomerContactsSql = `SELECT * FROM ${postgresConfig.getSchemaName()}.${customerContactTable}`;
    this.getSalesMappingsSql = `SELECT * FROM ${postgresConfig.getSchemaName()}.${salesMappingsTable}`;
    this.getConsignmentsSql = `SELECT * FROM ${postgresConfig.getSchemaName()}.${consignmentTable}`;
  }

  public search(postgresClient: PostgresClient) {
    return tryCatch(async (req, res, next) => {
      const text = req.body[getDataFields.searchText];
      const filters = req.body[getDataFields.filters];
        const client = await postgresClient.getClient()
        const results = await this.searchTable.search(text, filters, client)
          .finally(async () => {
            return postgresClient.release(client);
          });
        // console.log(`search results ${dump(this.searchTable.uniquify(results))}`)
        res.json(this.searchTable.uniquify(results));
    })
  }
  public getConsignments(postgresClient: PostgresClient) {
    return tryCatch(async(req, res, next) => {
        const customerId = req.params[getDataFields.customerId] || null;
        const client = await postgresClient.getClient()
        const results = await this.queryConsignments(client, customerId)
          .finally(async () => {
            console.log('trying to disconnect consignment')
            return postgresClient.release(client);
          });
        res.json(results.rows);
    })
  }
  public async queryConsignments(client, customerId: (string | null) = null) {
    let sql = this.getConsignmentsSql;
    if(customerId) {
      sql += ` WHERE customer_id='${customerId}'`;
    }
    sql +=';';
    return client.query(sql).catch(err => {
      throw new Error(`Mistake in '${sql}' caused ${err.message}`);
    });
  }
  public getCustomers(postgresClient: PostgresClient) {
    return tryCatch(async(req, res, next) => {
      const uuid = req.params[getDataFields.uuid] || null;
      const customerId = req.params[getDataFields.customerId] || null;
        const client = await postgresClient.getClient()
        const results = await queryRecords(client, this.getCustomersSql, 'customer_code',
          {uuid, id: customerId})
          .finally(async () => {
            console.log('trying to disconnect customer')
            return postgresClient.release(client);
          });
        res.json(results);
    })
  }
  public getCustomerContacts(postgresClient: PostgresClient) {
    return tryCatch(async(req, res, next) => {
      const uuid = req.params[getDataFields.uuid] || null;
      const customerId = req.params[getDataFields.customerId] || null;
      const client = await postgresClient.getClient()
      const results = await queryRecords(client, this.getCustomerContactsSql, 'ship_to_id',
        {uuid, id: customerId})
        .finally(async () => {
          console.log('trying to disconnect customer contacts')
          return postgresClient.release(client);
        });
      res.json(results);
    })
  }

  public getSalesMappings(postgresClient: PostgresClient) {
    return tryCatch(async(req, res, next) => {
      const uuid = req.params[getDataFields.uuid] || null;
      const customerId = req.params[getDataFields.customerId] || null;
      const client = await postgresClient.getClient()
      const results = await queryRecords(client, this.getSalesMappingsSql, 'customer_ship_to',
        {uuid, id: customerId})
        .finally(async () => {
          console.log('trying to disconnect customer contacts')
          return postgresClient.release(client);
        });
      res.json(results);
    })
  }
}
