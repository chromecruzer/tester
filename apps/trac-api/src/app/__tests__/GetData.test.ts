import GetData from "../GetData";
import {queryRecords} from '@trac/postgresql';

jest.mock('@trac/postgresql');
describe('GetData', () => {
  let getDataModel, mockClient, mockPostgresClient, postgresConfigFactory, searchResults, cleanedResults,
    mockQueryRecords;
  beforeEach(() => {
    mockQueryRecords = jest.mocked(queryRecords).mockReturnValue(Promise.resolve(['data1', 'data2']));
    searchResults = ['raw', 'version'];
    cleanedResults = ['cleaned', 'version'];
    postgresConfigFactory = {
      getSchemaName: () => 'schemanameFromConfig'
    }
    getDataModel = new GetData(postgresConfigFactory, 'consignmentTable',
      'customerTable');
    getDataModel.searchTable.search = jest.fn().mockReturnValue(Promise.resolve(searchResults));
    getDataModel.searchTable.uniquify = jest.fn().mockReturnValue(cleanedResults);
    mockClient = {
      query: jest.fn()
    }
    mockPostgresClient = {
      getClient: jest.fn().mockReturnValue(Promise.resolve(mockClient)),
      release: jest.fn()
    };
  });
  afterEach(() => {
    jest.resetAllMocks();
  });
  describe('search', () => {
    let searchFunc;
    beforeEach(() => {
      searchFunc = getDataModel.search(mockPostgresClient);
    });
    it('should call search table without filters correctly', () => {
      const res = {json: jest.fn()};
      return searchFunc({body: {searchText: 'text2Search'}}, res, jest.fn()).then(() => {
        expect(getDataModel.searchTable.search).toBeCalledWith('text2Search', undefined, mockClient);
        expect(getDataModel.searchTable.uniquify).toBeCalledWith(searchResults);
        expect(res.json).toBeCalledWith(cleanedResults);
      });
    });
    it('should call search table with filters correctly', () => {
      const res = {json: jest.fn()};
      return searchFunc({body: {searchText: 'text2Search', filters:{consignments: true, customers: true}}}, res, jest.fn()).then(() => {
        expect(getDataModel.searchTable.search).toBeCalledWith('text2Search',
          {consignments: true, customers: true}, mockClient);
        expect(getDataModel.searchTable.uniquify).toBeCalledWith(searchResults);
        expect(res.json).toBeCalledWith(cleanedResults);
      });
    });
    it('should create and release a client with success', () => {
      const res = {json: jest.fn()};
      return searchFunc({body: {searchText: 'text2Search'}}, res, jest.fn()).then(() => {
        expect(mockPostgresClient.getClient).toBeCalled();
        expect(mockPostgresClient.release).toBeCalledWith(mockClient);
      });
    });
    it('should create and release a client with failure', () => {
      getDataModel.searchTable.search = jest.fn().mockReturnValue(Promise.reject('i am not playing your game'));
      const res = {json: jest.fn()};
      return searchFunc({body: {searchText: 'text2Search'}}, res, jest.fn()).finally(() => {
        expect(mockPostgresClient.getClient).toBeCalled();
        expect(mockPostgresClient.release).toBeCalledWith(mockClient);
      });
    });
  });
  describe('getConsignments', () => {
    it('should query with the proper SQL when called with a customer ID', () => {
      const actualFunc = getDataModel.getConsignments(mockPostgresClient);
      mockClient.query.mockReturnValue(Promise.resolve({rows: ['data']}));
      return actualFunc({params: {customerId: 'idForCustomer'}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
      });
    });
    it('should query with the proper SQL when called without a customer ID', () => {
      const actualFunc = getDataModel.getConsignments(mockPostgresClient);
      mockClient.query.mockReturnValue(Promise.resolve({rows: ['data']}));
      return actualFunc({params: {}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockClient.query.mock.calls[0][0]).toMatchSnapshot();
      });
    });
  });
  describe('getCustomers', () => {

    it('should query with the proper SQL when called with a customer uuid', () => {
      const actualFunc = getDataModel.getCustomers(mockPostgresClient);
      mockClient.query.mockReturnValue(Promise.resolve({rows: ['data']}));
      return actualFunc({params: {uuid: 'uuidForCustomer'}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockQueryRecords.mock.calls).toMatchSnapshot();
      });
    });
    it('should query with the proper SQL when called without a customer uuid', () => {
      const actualFunc = getDataModel.getCustomers(mockPostgresClient);
      mockClient.query.mockReturnValue(Promise.resolve({rows: ['data']}));
      return actualFunc({params: {}}, {json: jest.fn()}, jest.fn()).then(() => {
        expect(mockQueryRecords.mock.calls).toMatchSnapshot();
      });
    });
  });
});
