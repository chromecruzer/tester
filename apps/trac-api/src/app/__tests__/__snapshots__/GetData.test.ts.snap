// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetData getConsignments should query with the proper SQL when called with a customer ID 1`] = `"SELECT * FROM schemanameFromConfig.consignmentTable WHERE customer_id='idForCustomer';"`;

exports[`GetData getConsignments should query with the proper SQL when called without a customer ID 1`] = `"SELECT * FROM schemanameFromConfig.consignmentTable;"`;

exports[`GetData getCustomers should query with the proper SQL when called with a customer uuid 1`] = `
Array [
  Array [
    Object {
      "query": [MockFunction],
    },
    "SELECT * FROM schemanameFromConfig.customerTable",
    "customer_code",
    Object {
      "id": null,
      "uuid": "uuidForCustomer",
    },
  ],
]
`;

exports[`GetData getCustomers should query with the proper SQL when called without a customer uuid 1`] = `
Array [
  Array [
    Object {
      "query": [MockFunction],
    },
    "SELECT * FROM schemanameFromConfig.customerTable",
    "customer_code",
    Object {
      "id": null,
      "uuid": null,
    },
  ],
]
`;
